api/requestresponse.go

```diff
@@ -57,8 +67,90 @@ func Redirect(w http.ResponseWriter, r *http.Request, url string, code int) {
 	http.Redirect(w, r, url, code)
 }
 
-// Decode should be used to convert the request's JSON body into the given v value.
-func Decode(w http.ResponseWriter, r *http.Request, v interface{}) error {
-	// FUTURE: Can handle different content types by looking at the request headers.
-	return json.NewDecoder(r.Body).Decode(v)
+// NotFound replies to the request with an HTTP 404 not found error, using the problem response defined by RFC 7807.
+func NotFound(w http.ResponseWriter, r *http.Request, extras ...ProblemExtra) {
+	code := http.StatusNotFound
+	title := http.StatusText(code)
+
+	Problem(w, r, title, title, code, extras...)
+}
+
+// Error replies to the request with the specified error message and HTTP code,
+// using the problem response defined by RFC 7807.
+func Error(w http.ResponseWriter, r *http.Request, err string, code int, extras ...ProblemExtra) {
+	Problem(w, r, http.StatusText(code), err, code, extras...)
+}
+
+//
+// Problem
+//
+
+// problem is a wrapper around a map that is used to collect all fields required for a problem response in
+// accordance to RFC 7807. It implements the json Marshaler interface to define a custom json marshalling.
+type problem struct {
+	fields map[string]interface{}
+}
+
+// MarshalJSON implements the json Marshaler interface.
+func (p problem) MarshalJSON() ([]byte, error) {
+	// Use the default marshaller for nested fields map.
+	return json.Marshal(p.fields)
+}
+
+// ProblemExtra provides a way to add extra information into the problem response.
+type ProblemExtra func(*problem)
+
+// WithType allows the type field to be added to the problem response.
+func WithType(t string) ProblemExtra {
```

Comment:
The most ridiculous of nitpicks, but you use full words `detail` and `instance` below, but a single char `t` here

Reply:
🙈 I'll update to single chars everywhere, 'type' is a keyword.